src/coreclr/jit/codegenxarch.cpp

```diff
@@ -2825,6 +2834,12 @@ void CodeGen::genCodeForInitBlkUnroll(GenTreeBlk* node)
                 regSize = 8;
             }
 #endif
+            if (regSize == YMM_REGSIZE_BYTES && bytesWritten + regSize > size)
+            {
+                // Step down from YMM registers to XMM registers
```

Comment:
For the init block, there is sometimes a byte pattern (I think, [this comment suggests that's what's going on](https://github.com/dotnet/runtime/blob/d05950ba1ad440855f9851857bbad6f9d1a34321/src/coreclr/jit/lowerxarch.cpp#L215-L220)) that is being written instead of just copying bytes from one address+offset to another address+offset. If we use the exact same shift back technique as the copy block, I think we can run into cases where the byte pattern won't line up correctly. I decided it would be easier to step down to an XMM register, then fall out of the loop instead of shifting back to avoid potentially clobbering what was previously written.

Let me know if my understanding of this case is incorrect, specifically about the byte pattern being used during the init block. I don't expect a performance difference between stepping down the word size vs shifting back, since it will be the same SIMD register, the only thing that changes is the word size.

Reply:
The main advantage of the shift-back technique for the copy block is to avoid allocating/using the GPR, but I don't think we'll be able to avoid using the GPR in the init block if the size doesn't perfectly fit in a SIMD register.

Reply:
>  think we can run into cases where the byte pattern won't line up correctly.

So the initblock can have only a byte pattern, so it means that each byte in our SIMD init register is the same, so each byte in the dst should be the same value. If we had a 2-byte pattern then we could have a problem, but when our pattern size is the same as the minimal addressable size I don't see how we can hit an issue.

> I don't expect a performance difference between stepping down the word size vs shifting back, since it will be the same SIMD register, the only thing that changes is the word size.

I am expecting a performance difference for cases like: init block with size 31. In the current version we will have:
[0, 16), [16, 24), [24, 28), [28, 30), [30, 31),
I am proposing to have
[0, 16), [15, 31).